### ISSUE TITLE
Only C op is implemented

### ISSUE THREAD

**Issue by LucaNap** [rodluger/starry#304]
Hello, 

I am trying to understand what this error means and how can it be solved:

`---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
~\AppData\Local\Temp\ipykernel_5436\751011257.py in <module>
     34     sys = starry.System(A, b)
     35     # Our model for the flux
---> 36     flux_model = pm.Deterministic("flux_model", sys.flux(t))
     37 
     38     # This is how we tell `pymc3` about our observations we are assuming they are ampally distributed about

C:\Miniconda3\envs\starry\lib\site-packages\starry\kepler.py in flux(self, t, total, integrated)
   1013                 from each body.
   1014         """
-> 1015         X = self.design_matrix(t)
   1016 
   1017         # Weight the ylms by amplitude

C:\Miniconda3\envs\starry\lib\site-packages\starry\kepler.py in design_matrix(self, t)
    998                         sec._map, "_sigr", self._math.to_array_or_tensor(0.0)
    999                     )
-> 1000                     for sec in self._secondaries
   1001                 ]
   1002             ),

C:\Miniconda3\envs\starry\lib\site-packages\starry\_core\utils.py in wrapper(instance, *args)
    139 
    140             # Just return the function as is
--> 141             return func(instance, *args)
    142 
    143         else:

C:\Miniconda3\envs\starry\lib\site-packages\starry\_core\core.py in X(self, t, pri_r, pri_m, pri_prot, pri_t0, pri_theta0, pri_amp, pri_inc, pri_obl, pri_fproj, pri_u, pri_f, sec_r, sec_m, sec_prot, sec_t0, sec_theta0, sec_porb, sec_ecc, sec_w, sec_Omega, sec_iorb, sec_amp, sec_inc, sec_obl, sec_u, sec_f, sec_sigr)
   2462                 pri_obl,
   2463                 pri_u,
-> 2464                 pri_f,
   2465             )
   2466         if self._reflected:

C:\Miniconda3\envs\starry\lib\site-packages\starry\_core\utils.py in wrapper(instance, *args)
    139 
    140             # Just return the function as is
--> 141             return func(instance, *args)
    142 
    143         else:

C:\Miniconda3\envs\starry\lib\site-packages\starry\_core\core.py in X(self, theta, xo, yo, zo, ro, inc, obl, u, f)
    782 
    783         """
--> 784         flux = self.flux(xo, yo, zo, ro, u)
    785         X = tt.reshape(flux, (-1, 1))
    786         return X

C:\Miniconda3\envs\starry\lib\site-packages\starry\_core\utils.py in wrapper(instance, *args)
    139 
    140             # Just return the function as is
--> 141             return func(instance, *args)
    142 
    143         else:

C:\Miniconda3\envs\starry\lib\site-packages\starry\_core\core.py in flux(self, xo, yo, zo, ro, u)
    759 
    760         # Get the Agol `c` coefficients
--> 761         c = self._get_cl(u)
    762         if self.udeg == 0:
    763             c_norm = c / (np.pi * c[0])

C:\Miniconda3\envs\starry\lib\site-packages\theano\graph\op.py in __call__(self, *inputs, **kwargs)
    251 
    252         if config.compute_test_value != "off":
--> 253             compute_test_value(node)
    254 
    255         if self.default_output is not None:

C:\Miniconda3\envs\starry\lib\site-packages\theano\graph\op.py in compute_test_value(node)
    128     thunk.outputs = [storage_map[v] for v in node.outputs]
    129 
--> 130     required = thunk()
    131     assert not required  # We provided all inputs
    132 

C:\Miniconda3\envs\starry\lib\site-packages\theano\graph\op.py in rval(p, i, o, n)
    474             # default arguments are stored in the closure of `rval`
    475             def rval(p=p, i=node_input_storage, o=node_output_storage, n=node):
--> 476                 r = p(n, [x[0] for x in i], o)
    477                 for o in node.outputs:
    478                     compute_map[o][0] = True

C:\Miniconda3\envs\starry\lib\site-packages\starry\_core\ops\limbdark\base_op.py in perform(self, *args)
     50 
     51     def perform(self, *args):
---> 52         raise NotImplementedError("Only C op is implemented")

NotImplementedError: Only C op is implemented`

Any help is appreciated!

**Comment by dfm** [comment]
Great - that's more like something we can handle! It looks like you'll want to add `-fexceptions` as a compile flag. You can probably do something like:

```python
import theano

theano.config.gcc__cxxflags += " -fexceptions"
```

Probably best to do that before importing starry.

**Comment by rodluger** [comment]
Looks like the error message hasn't changed:
```
error: exception handling disabled, use \'-fexceptions\' to enable.
```
@dfm is right about the cause. I'll try to think of a workaround.

**Comment by dfm** [comment]
Yeah - if you look at the full error linked above you'll see that the `-fno-exceptions` flag is being set somewhere.

**Comment by dfm** [comment]
It may be sensible to remove this exception from starry anyways since it would result in a segfault, I think!

**Comment by rodluger** [comment]
Yeah. The exceptions date back to when we were using PyBind11, I think, which caught them and raised corresponding python exceptions. @dfm what's the proper way of raising an error in the c code of a theano op? Return some special flag and process it on the python side?

**Comment by dfm** [comment]
In this case I think we can just remove the exceptions since they shouldn't ever be hit given that we call it very carefully ;)

In the more general case, something like your suggestion @rodluger — handling the error in Python or in CPython [here](https://github.com/rodluger/starry/blob/da7fee48c5ef94278f0047be0579e2f13492cdd5/starry/_core/ops/limbdark/limbdark.cc) — but in this case it might be more trouble than it's worth to propagate the error all the way up manually.

**Comment by LucaNap** [comment]
@rodluger , yes it works now! Thank you

Note that only the following order of setup worked (while others did raise errors, for example when I did not install the older version of starry first).

> conda create --name starry python==3.9.7 mkl-service -y
> conda activate starry
> pip install numpy==1.21.0 starry tqdm corner jupyterlab ipykernel packaging lightkurve
> pip install git+https://github.com/rodluger/starry

**Comment by rodluger** [comment]
Glad to hear. Unfortunately `starry` is getting harder and harder for me to support as `theano` falls apart with newer versions of `numpy`, `pymc`, `gcc`, etc. So these issues are likely to pop up again. @dfm and I are talking about rewriting `starry` in `jax` and integrating it directly into `exoplanet` -- I think that is the future of this code.
